### DISCUSSION TITLE
Working with Jenkins

### DISCUSSION THREAD

**Discussion by kmayer10** [kmayer10/bfl-interns#4] [Show and tell]
#### Installing `JDK`
```
sudo apt-get update && sudo apt-get install -y default-jdk
```

**Comment by kmayer10** [comment]
#### Download Jenkins War
```
https://get.jenkins.io/war/2.389/jenkins.war
```
#### Starting Jenkins
```
nohup java -jar jenkins.war &
```
#### Open Port `8080` in Azure Network Security Group

**Reply by kmayer10** [reply]
#### try to access application using http://your-server-public-ip:8080
![image](https://user-images.githubusercontent.com/15359375/216888341-1b4ce02f-dcf6-4aa3-9c12-5b2e0be30275.png)

**Comment by kmayer10** [comment]
![image](https://user-images.githubusercontent.com/15359375/216889085-528cf0de-d7e4-4c28-a540-90b1197ec2e2.png)
![image](https://user-images.githubusercontent.com/15359375/216889151-23be28d8-99b8-4f84-be09-9671fbf36163.png)
![image](https://user-images.githubusercontent.com/15359375/216889384-8c6d899f-73f2-487c-b55a-0e35fb7395ac.png)
![image](https://user-images.githubusercontent.com/15359375/216889444-644c604a-7a55-483f-9b1d-706a3e6a5254.png)
![image](https://user-images.githubusercontent.com/15359375/216889477-97a70a68-839c-43e8-8d10-cdbb00d27c44.png)
![image](https://user-images.githubusercontent.com/15359375/216889514-d71121cf-829e-4201-aaa2-ba8e78926d7e.png)

**Comment by kmayer10** [comment]
![image](https://user-images.githubusercontent.com/15359375/216890824-055fe934-914c-4f7b-9431-7ab0b9cf8e57.png)
![image](https://user-images.githubusercontent.com/15359375/216890899-84d89772-25c0-457e-b8f1-ff3ace716b4c.png)
![image](https://user-images.githubusercontent.com/15359375/216890960-800fbc33-bd3e-42d1-9b8c-9f5e00b21cb1.png)
![image](https://user-images.githubusercontent.com/15359375/216891010-6bbcde7e-c535-4b8e-b95d-4288252aec57.png)
![image](https://user-images.githubusercontent.com/15359375/216891049-4e494d4c-03b4-40ab-bfe3-a71a86aaea6c.png)
![image](https://user-images.githubusercontent.com/15359375/216891598-1d852c5d-96dd-4604-a13c-e2f29478332d.png)
#### Run `cat ~/.ssh/id_rsa` to print and copy the private key to be added to Jenkins Global Credentials.

**Comment by kmayer10** [comment]
![image](https://user-images.githubusercontent.com/15359375/216896302-2aba32c3-980c-49b4-a408-4e5f4deb3c50.png)
![image](https://user-images.githubusercontent.com/15359375/216896422-80a67bf7-39d5-4aa3-80b9-2f8b753bcc2b.png)
![image](https://user-images.githubusercontent.com/15359375/216896755-04b8ad74-531b-4094-b6bf-218eb1049190.png)
#### Go to Build Steps > Execute Shell and add the below lines
```
#!/bin/bash
echo "[INFO] Hi from `hostname` used as Jenkins Host"
echo "[INFO] My Current Directory is:"
pwd
echo "[INFO] Current time is:"
date
```
![image](https://user-images.githubusercontent.com/15359375/216897101-ce40f1c5-374c-46d7-a58e-c4130e173d20.png)
![image](https://user-images.githubusercontent.com/15359375/216897169-de6294ef-befa-490e-a996-7ac1fd020ada.png)

**Comment by kmayer10** [comment]
![image](https://user-images.githubusercontent.com/15359375/216898093-9efef8b8-a197-425a-bf6a-083d9ecc0068.png)
![image](https://user-images.githubusercontent.com/15359375/216898161-cd6efb37-ef48-4359-b324-764cc3a68b3f.png)
#### Click on `Console Output`
![image](https://user-images.githubusercontent.com/15359375/216898347-ca5a4701-ea86-4fca-b239-ddff3532ea4b.png)

**Comment by kmayer10** [comment]
#### Create new build using `New Item` > Select Freestyle option and add the below code after selecting `Execute Shell` as Build Step
```
#!/bin/bash
echo "[INFO] Current working directory is: `pwd`"
ls -ltr
./run-script.sh
```
![image](https://user-images.githubusercontent.com/15359375/216899888-73e9849f-92fd-44ed-8a12-a658fbc3fefa.png)
#### `Save` & Click `Build Now`
![image](https://user-images.githubusercontent.com/15359375/216900021-fca33b48-30b9-4026-9f75-44e60d98fc3c.png)
![image](https://user-images.githubusercontent.com/15359375/216900264-2640586c-04fa-42c6-9dcb-2cabf21514ff.png)

**Comment by kmayer10** [comment]
```
git@github.com:kmayer10/bfl-interns.git
```

**Comment by natashanangia1825** [comment]
#### command to add github verification key in known hosts file
```
ssh-keyscan github.com >> ~/.ssh/known_hosts
```

#### expected output
```
# github.com:22 SSH-2.0-babeld-fa280caa
# github.com:22 SSH-2.0-babeld-fa280caa
# github.com:22 SSH-2.0-babeld-fa280caa
# github.com:22 SSH-2.0-babeld-fa280caa
# github.com:22 SSH-2.0-babeld-fa280caa
```

**Comment by kmayer10** [comment]
#### add a new file with the name `run-script.sh` in your repository with the below code
```
#!/bin/bash
echo "[INFO] Runing from Script added just now"
echo "[INFO] Current time is:"
date
```
#### Please ensure the file has `755` as permission
```
-rwxr-xr-x  1 kmayer kmayer    97 Feb  6 07:24 run-script.sh*
```
#### run `chmod 755 run-script.sh" to change the permissions to `755`

**Comment by kmayer10** [comment]
![image](https://user-images.githubusercontent.com/15359375/216913020-499d1be4-cf0e-48b6-8087-d6ad2481572b.png)
![image](https://user-images.githubusercontent.com/15359375/216913494-03b175ea-45da-4a3a-bbb7-cb3d70d133fe.png)
#### Entry is `* * * * *` & `Save`
#### Added some changes in the configured repository and wait for a min to check if build will get auto triggered

**Comment by kmayer10** [comment]
![image](https://user-images.githubusercontent.com/15359375/216916017-1b77b299-0511-4e3d-9216-bf3acebe7769.png)
![image](https://user-images.githubusercontent.com/15359375/216916174-9afc3b20-8850-4742-9523-ab37cd20ea22.png)
![image](https://user-images.githubusercontent.com/15359375/216916291-229c1cc5-27ae-4bea-a933-cac87316881d.png)
![image](https://user-images.githubusercontent.com/15359375/216916566-8a5f42c8-1518-40da-abd0-8aae3a8d8ca9.png)
#### Add Jenkins URL as below format
```
http://20.106.212.241:8080/github-webhook/
```
![image](https://user-images.githubusercontent.com/15359375/216916837-de41cedf-f437-4181-bafc-52a66ae320be.png)

**Comment by kmayer10** [comment]
#### fork https://github.com/kmayer10/maven-sample-project in your account

**Comment by kmayer10** [comment]
![image](https://user-images.githubusercontent.com/15359375/216935218-d83f3e35-f160-47ec-af21-030227397aea.png)

**Comment by kmayer10** [comment]
![image](https://user-images.githubusercontent.com/15359375/216936282-abd3d22f-58bb-4851-8411-82450d4f4fad.png)
![image](https://user-images.githubusercontent.com/15359375/216936538-73cfb95b-68be-470f-a52d-3686be885542.png)
![image](https://user-images.githubusercontent.com/15359375/216936949-be9c476d-df42-4e04-bc22-e0c8fa4b17c5.png)

**Comment by kmayer10** [comment]
![image](https://user-images.githubusercontent.com/15359375/216939551-8c31ce7c-20f8-47a9-9e1a-48d0198af283.png)
#### Run the build again post successfully updating the changes as suggested in the above image
#### file to be archived is `target/devops.war`
![image](https://user-images.githubusercontent.com/15359375/216939678-3a14815c-aab1-437f-a2d4-906f01d5f838.png)

**Comment by kmayer10** [comment]
#### Go to Dashboard > Manage Jenkins > Manage Plugins > Available Plugins > Search for `pipeline utility steps`
![image](https://user-images.githubusercontent.com/15359375/216942292-7fb3e777-9674-44f9-b69c-40b79874e16b.png)
#### Click `Install without restart`
![image](https://user-images.githubusercontent.com/15359375/216942424-c4740c49-4b29-48a4-8363-2f4747f50257.png)

**Comment by kmayer10** [comment]
![image](https://user-images.githubusercontent.com/15359375/216942935-96cb03ad-3de3-4c2b-aaea-254b058f4851.png)
![image](https://user-images.githubusercontent.com/15359375/216943520-fb530aa6-e244-4a6b-9ab5-0fa572f54d4c.png)
#### under pipeline script add the below code & click `Save`
```
pipeline {
    agent any
    tools {
        maven "maven_3"
    }
    stages {
        stage('Build') {
            steps {
                // Get some code from a GitHub repository
                git 'git@github.com:kmayer10/maven-sample-project.git'
                // Run Maven on a Unix agent.
                sh "mvn -Dmaven.test.failure.ignore=true clean package"
            }

            post {
                // If Maven was able to run the tests, even if some of the test
                // failed, record the test results and archive the jar file.
                success {
                    archiveArtifacts 'target/devops.war'
                }
            }
        }
    }
}
```
#### Click `Build Now`
![image](https://user-images.githubusercontent.com/15359375/216943967-e6554c62-bee8-4a3b-88b5-737ba1f50591.png)

**Comment by kmayer10** [comment]
![image](https://user-images.githubusercontent.com/15359375/216944978-62634f92-2b39-4ec1-9b59-7ac4fb542089.png)
![image](https://user-images.githubusercontent.com/15359375/216945472-bfe2e471-55c3-40fe-afd6-1cfda2c62588.png)

**Comment by kmayer10** [comment]
![image](https://user-images.githubusercontent.com/15359375/216946518-6e0a6e59-eb70-413b-8e7c-ded6359e34b8.png)
![image](https://user-images.githubusercontent.com/15359375/216946745-a7c0ae6f-9320-4bfe-8d44-cddbc776c755.png)
![image](https://user-images.githubusercontent.com/15359375/216947024-c2568361-028a-4467-9b59-0e7e244df1a5.png)
